### PR TITLE
Fix for button click tracking not working

### LINKED ISSUE #10323: Mautic does not track MJML button clicks (mj-button)
## Bug Description
Button links in emails (segment and template) with an https://documentation.mjml.io/#mj-button tag do not get tracked in Mautic. 

Normal text links and opens are tracked.

| Q   | A
| --- | ---
| Mautic version | 4.0 RC
| PHP version | 7.4 / DDEV
| Browser | Chrome

### Steps to reproduce
1.  Create a new email with the "Confirm Me" theme
2. Add a link to the button
3. Add another link to any of the text
4. Send the email
5. Check the link. It should point to the Mautic domain, not the original target
 
Text link gets converted
```html
<p>
  <a
    class="link"
    href="http://mautic.ddev.site/index_dev.php/r/d01a7f66566927f9a20a1ec0f?ct=YTo0OntzOjY6InNvdXJjZSI7YTowOnt9czo1OiJlbWFpbCI7TjtzOjQ6InN0YXQiO3M6MjI6IjYxMTI3MDA4ZGQ0NDgzNTEyMjEzNTkiO3M6NDoibGVhZCI7czoxOiIxIjt9"
    >Company</a
  >
  Name<br />
  111 Amazing Street<br />
  Beautiful City
</p>
```

Button link stay the same
```html
<table
  border="0"
  cellpadding="0"
  cellspacing="0"
  role="presentation"
  style="border-collapse: separate; line-height: 100%"
>
  <tr>
    <td
      align="center"
      bgcolor="#F45E43"
      role="presentation"
      style="
        border: none;
        border-radius: 3px;
        cursor: auto;
        font-style: bold;
        mso-padding-alt: 10px 25px;
        background: #f45e43;
      "
      valign="middle"
    >
      <a
        href="https://aivie.ch"
        style="
          display: inline-block;
          background: #f45e43;
          color: #ffffff;
          font-family: Ubuntu, Helvetica, Arial, sans-serif;
          font-size: 13px;
          font-style: bold;
          font-weight: normal;
          line-height: 120%;
          margin: 0;
          text-decoration: none;
          text-transform: none;
          padding: 10px 25px;
          mso-padding-alt: 0px;
          border-radius: 3px;
        "
        target="_blank"
      >
        Yes, subscribe me to the list
      </a>
    </td>
  </tr>
</table>
```

### PR DESCRIPTION
**Pull request by adiux**
fixes: https://github.com/mautic/mautic/issues/10323

### CODE REVIEW

**Comment by adiux** [comment]
https://github.com/mautic/mautic/issues/10331

File: dist/commands.js [hunk 1]
```diff
@@ -5,19 +5,22 @@ export default (editor => {
...
-  }); // Slot to {token}
+    stop: () => dynamicContentCmd.stopDynamicContentPopup()
+  }); // Component to {token}
 
-  editor.Commands.add('preset-mautic:dynamic-content-slots-to-tokens', {
-    run: edtr => dynamicContentCmd.convertDynamicContentSlotsToTokens(edtr)
-  }); // {token} to slot
+  editor.Commands.add('preset-mautic:dynamic-content-components-to-tokens', {
+    run: edtr => dynamicContentCmd.convertDynamicContentComponentsToTokens(edtr)
+  }); // Link store to compoennt
```
Comment (RCheesley) [inline dist/commands.js]:
```suggestion
  }); // Link store to component
```

File: dist/dynamicContent/dynamicContent.commands.js [hunk 2]
```diff
@@ -49,33 +47,44 @@ export default class DynamicContentCommands {
...
-        const dynConId = DynamicContentCommands.getDcStoreId(attributes['data-param-dec-id']);
-        const dynConTarget = mQuery(dynConId);
-        const dynConName = dynConTarget.find(`${dynConId}_tokenName`).val();
-        const dynConToken = `{dynamiccontent="${dynConName}"}`; // Clear id because it's reloaded by Mautic and this prevent slot to be destroyed by GrapesJs destroy event on close.
-        // dynamicContent.addAttributes({ 'data-param-dec-id': '' });
+      const decId = attributes['data-param-dec-id'];
 
-        dynamicContent.set('content', dynConToken);
+      if (!decId) {
+        this.logger.debug('Expected a dynamic content component', {
```
Comment (RCheesley) [inline dist/dynamicContent/dynamicContent.commands.js]:
```suggestion
        this.logger.debug('Expected a Dynamic Content component', {
```

File: dist/dynamicContent/dynamicContent.commands.js [hunk 3]
```diff
@@ -49,33 +47,44 @@ export default class DynamicContentCommands {
...
+          dynamicContent
+        });
+        throw new Error('no dynamic content component');
```
Comment (RCheesley) [inline dist/dynamicContent/dynamicContent.commands.js]:
```suggestion
        throw new Error('No Dynamic Content component');
```

File: dist/dynamicContent/dynamicContent.service.js [hunk 4]
```diff
@@ -113,23 +204,25 @@ export default class DynamicContentService {
...
     component.addAttributes({
-      'data-param-dec-id': parseInt(dcTarget.replace(/[^0-9]/g, ''), 10)
+      'data-param-dec-id': decId
     }); // Replace token on canvas with user facing name: dcName
 
-    component.set('content', dcName);
-    this.logger.debug('Created a new dynamic content item', {
-      dcName,
+    component.set('content', `Dynamic Content ${decId}`);
+    this.logger.debug('Created a new dynamic content item in store', {
```
Comment (RCheesley) [inline dist/dynamicContent/dynamicContent.service.js]:
```suggestion
    this.logger.debug('Created a new Dynamic Content item in store', {
```

File: dist/mjml/mjml.service.js [hunk 5]
```diff
@@ -46,12 +49,14 @@ export default class MjmlService {
 
   static mjmlToHtml(mjml) {
     try {
-      if (typeof mjml !== 'string') {
+      if (typeof mjml !== 'string' || !mjml.includes('<mjml>')) {
         throw new Error('no valid mjml string');
```
Comment (RCheesley) [inline dist/mjml/mjml.service.js]:
```suggestion
        throw new Error('No valid MJML string');
```

File: dist/buttons/buttonClose.command.js [hunk 6]
```diff
@@ -18,7 +18,7 @@ export default class ButtonCloseCommands {
       throw new Error('no email-html editor');
```
Comment (RCheesley) [inline dist/buttons/buttonClose.command.js]:
```suggestion
      throw new Error('No email-HTML editor');
```

File: dist/buttons/buttonClose.command.js [hunk 7]
```diff
@@ -31,7 +31,7 @@ export default class ButtonCloseCommands {
       throw new Error('no email-mjml editor');
```
Comment (RCheesley) [inline dist/buttons/buttonClose.command.js]:
```suggestion
      throw new Error('No email-MJML editor');
```

File: dist/dynamicContent/dynamicContent.commands.js [hunk 8]
```diff
@@ -49,33 +47,44 @@ export default class DynamicContentCommands {
...
       }
+
+      const dynConId = DynamicContentCommands.getDcStoreId(attributes['data-param-dec-id']);
+      const dynConTarget = mQuery(dynConId);
+      const dynConName = dynConTarget.find(`${dynConId}_tokenName`).val();
+      const dynConToken = `{dynamiccontent="${dynConName}"}`; // Clear id because it's reloaded by Mautic and this prevent slot to be destroyed by GrapesJs destroy event on close.
+      // dynamicContent.addAttributes({ 'data-param-dec-id': '' });
+
+      this.logger.debug('Replaced components content with its token', {
```
Comment (RCheesley) [inline dist/dynamicContent/dynamicContent.commands.js]:
```suggestion
      this.logger.debug('Replaced component's content with its token', {
```

**Review by RCheesley — DISMISSED** [review]
Superficial review for grammar etc.

A few suggestions to make sure the messages we show to users/output to logs are consistent with our style guide. Main things to look for (as some I could not update with this review) are:

- Capitalise the first word
- Capitalise any Mautic features, for example Dynamic Content, and ensure they are written out fully and correctly (a few places have used DynamicContent without a space)
- Capitalise any abbreviations such as HTML and MJML

**Review by dennisameling — APPROVED** [review]
Looks good to me! Appreciate the comments in the code so that it's easier for future contributors to see what's going on 👍🏼

**Pull request merged**